src/bio/bio_dev_monitor.c

```diff
@@ -0,0 +1,442 @@
+		"WARNING" : "OK");
+	D_PRINT("  Temperature:               %s\n",
+		health_page->critical_warning.bits.temperature ?
+		"WARNING" : "OK");
+	D_PRINT("  Device Reliability:        %s\n",
+		health_page->critical_warning.bits.device_reliability ?
+		"WARNING" : "OK");
+	D_PRINT("  Read Only:                 %s\n",
+		health_page->critical_warning.bits.read_only ? "Yes" : "No");
+	D_PRINT("  Volatile Memory Backup:    %s\n",
+		health_page->critical_warning.bits.volatile_memory_backup ?
+		"WARNING" : "OK");
+	D_PRINT("  Current Temperature:       %u Kelvin (%d Celsius)\n",
+		health_page->temperature, (int)health_page->temperature - 273);
+	D_PRINT("Available Spare:             %u%%\n",
+		health_page->available_spare);
+	D_PRINT("Available Spare Threshold:   %u%%\n",
+		health_page->available_spare_threshold);
+	D_PRINT("Life Percentage Used:        %u%%\n",
+		health_page->percentage_used);
+	D_PRINT("Data Units Read:             ");
+	dprint_uint128_dec(health_page->data_units_read);
+	D_PRINT("\n");
+	D_PRINT("Data Units Written:          ");
+	dprint_uint128_dec(health_page->data_units_written);
+	D_PRINT("\n");
+	D_PRINT("Host Read Commands:          ");
+	dprint_uint128_dec(health_page->host_read_commands);
+	D_PRINT("\n");
+	D_PRINT("Host Write Commands:         ");
+	dprint_uint128_dec(health_page->host_write_commands);
+	D_PRINT("\n");
+	D_PRINT("Controller Busy Time:        ");
+	dprint_uint128_dec(health_page->controller_busy_time);
+	D_PRINT(" minutes\n");
+	D_PRINT("Power Cycles:                ");
+	dprint_uint128_dec(health_page->power_cycles);
+	D_PRINT("\n");
+	D_PRINT("Power On Hours:              ");
+	dprint_uint128_dec(health_page->power_on_hours);
+	D_PRINT(" hours\n");
+	D_PRINT("Unsafe Shutdowns:	     ");
+	dprint_uint128_dec(health_page->unsafe_shutdowns);
+	D_PRINT("\n");
+	D_PRINT("Unrecoverable Media Errors:  ");
+	dprint_uint128_dec(health_page->media_errors);
+	D_PRINT("\n");
+	D_PRINT("Lifetime Error Log Entries:  ");
+	dprint_uint128_dec(health_page->num_error_info_log_entries);
+	D_PRINT("\n");
+	D_PRINT("Warning Temperature Time:    %u minutes\n",
+		health_page->warning_temp_time);
+	D_PRINT("Critical Temperature Time:   %u minutes\n",
+		health_page->critical_temp_time);
+
+	/* Prep NVMe command to get controller data */
+	ctrlr_page_sz = sizeof(struct spdk_nvme_ctrlr_data);
+	memset(&cmd, 0, sizeof(cmd));
+	cmd.opc = SPDK_NVME_OPC_IDENTIFY;
+	cmd.cdw10 = SPDK_NVME_IDENTIFY_CTRLR;
+
+	/*
+	 * Submit an NVMe Admin command to get controller data
+	 * to the bdev.
+	 */
+	rc = spdk_bdev_nvme_admin_passthru(health_desc, channel, &cmd,
+					   dev_health->bhm_ctrlr_buf,
+					   ctrlr_page_sz,
+					   get_spdk_identify_ctrlr_completion,
+					   dev_health);
+	if (rc)
+		D_ERROR("NVMe admin passthru (identify ctrlr), rc:%d\n", rc);
+}
+
+/* Get the SPDK device health state log and print all useful stats */
+void
+query_spdk_health_stats(struct bio_xs_context *ctxt, uint64_t now)
```

Comment:
query_health_stats() or bio_bs_monitor()? 'spdk' shouldn't appear in this function name.